xarray/core/datatree.py

```diff
@@ -486,6 +486,35 @@ def _pre_attach(self: DataTree, parent: DataTree, name: str) -> None:
         parent_ds = parent._to_dataset_view(rebuild_dims=False, inherited=True)
         check_alignment(path, node_ds, parent_ds, self.children)
 
+    def _dedup_inherited_coordinates(self):
+        # This method removes repeated indexes (and correpsonding coordinates)
+        # that are repeated between a DataTree and its parents.
+        #
+        # TODO(shoyer): Decide how to handle repeated coordinates *without* an
+        # index. Should these be allowed, in which case we probably want to
+        # exclude them from inheritance, or should they be automatically
+        # dropped?
+        # https://github.com/pydata/xarray/issues/9475#issuecomment-2357004264
+        removed_something = False
+        for name in self._indexes.parents:
+            if name in self._node_indexes:
+                # Indexes on a Dataset always have a corresponding coordinate
+                del self._node_indexes[name]
+                del self._node_coord_variables[name]
+                removed_something = True
```

Comment:
So the answer to https://github.com/pydata/xarray/pull/9510#discussion_r1763445717 is that here if two indexes have the same name on parent and child then they must be equal (i.e. duplicated), because if they weren't then the alignment check would have failed? If so I think that deserves a comment in the code.

Reply:
I also wonder if that means we can do the deduplication in or near to the same method that we use to check alignment...?

Reply:
I added another clarifying comment.

It looks like moving it to `_pre_attach()` works, which is definitely a bit clearer.